### ISSUE TITLE
Ability to check from inside the function in which process it's called (fixed/idle)

### ISSUE THREAD

**Issue by kubecz3k** [godotengine/godot#3698]
Some functions don't work properly when they are called from idle process (for example when working with PhysicServer). It would be good to have ability to check in which process you are.

**Comment by bojidar-bg** [comment]
I think this would be impossible to do if running _fixed_process within a thread, since both _process and _fixed_process might occur at once in this case...

**Comment by akien-mga** [comment]
@RandomShaper did a nice autopsy ;) of the way fixed_process and idle work, he might have an idea.

**Comment by reduz** [comment]
Those should never run in a thread..

**Comment by kubecz3k** [comment]
@bojidar-bg @reduz I think I was not clear enough. This issue was never about custom threads. I have some static helper functions in my project 'Helpers' singleton, some of them are used to check directly in physic server if some bodies overlaps or if some body is colliding with ray (very handy for example at object spawn time). Sometimes I'm forgetting that those methods are not working in idle process, but it's impossible for me to write some kind of assertion/warning for this currently.

**Comment by reduz** [comment]
It is possible to check, I think there is a function somewhere. i know
InputDefault uses that function for just pressed

On Apr 26, 2017 10:45 PM, "kubecz3k" <notifications@github.com> wrote:

@bojidar-bg <https://github.com/bojidar-bg> @reduz
<https://github.com/reduz> I think I was not clear enough. This issue was
never about custom threads. I have some static helper functions in my
project 'Helpers' singleton, some of them are used to check directly in
physic server if some bodies overlaps or if some body is colliding with ray
(very handy for example at object spawn time). Sometimes I'm forgetting
that those methods are not working in idle process, but it's impossible for
me to write some kind of assertion/warning for this currently.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/godotengine/godot/issues/3698#issuecomment-297535809>,
or mute the thread

**Comment by kubecz3k** [comment]
Oh I assume it's about those lines: 
https://github.com/godotengine/godot/blob/master/main/input_default.cpp#L159
```
if (Engine::get_singleton()->is_in_fixed_frame()) {
   return E->get().pressed && E->get().fixed_frame == Engine::get_singleton()->get_fixed_frames();
} else {
   return E->get().pressed && E->get().idle_frame == Engine::get_singleton()->get_idle_frames();
}
```

**Comment by kubecz3k** [comment]
It seems its not exposed to GdScript, so it seems all that need to be done is to expose them :) Thanks @reduz

**Comment by RandomShaper** [comment]
A possible workaround would be catching the 'fixed_frame' and 'idle_frame' signals emitted by the SceneTree.

After each, you know you are in the named kind of process until another one comes. Note that you may get multiple fixed frame signals as multiple fixed frames can happen per idle frame.

Anyway, if you are going to expose the relevant method, you don't need to do that. But just for the records.

**Comment by kubecz3k** [comment]
solved #9155
